### ISSUE TITLE
Parsing error related to grammar nonterminals

### ISSUE THREAD

**Issue by polgreen** [CVC4/CVC4#4024]
**Describe the bug**

CVC4 fails to parse the example on the SyGuS website where the grammar contains a line which specifies the nonterminals: https://sygus.org/language/
~~~
(set-logic LIA)
(synth-fun max2 ((x Int) (y Int)) Int
    ((I Int) (B Bool))
    ((I Int (x y 0 1
             (+ I I) (- I I)
             (ite B I I)))
     (B Bool ((and B B) (or B B) (not B)
              (= I I) (<= I I) (>= I I))))
)

(declare-var x Int)
(declare-var y Int)
(constraint (>= (max2 x y) x))
(constraint (>= (max2 x y) y))
(constraint (or (= x (max2 x y)) (= y (max2 x y))))
(check-synth)
~~~

Error message:
~~~
CVC4 Error:
Parse Error: test2.sl:8.14: Expected <invalid>, found ')'.

      ((I Int) (B Bool))
~~~

**Command line arguments**: no command line arguments
**CVC4 version/commit**: CVC4 version 1.8-prerelease [git master bcaebfa1]
**Operating system**: macOS 10.15.3

**`configure.sh` options**
Paste `configure.sh` here.

**`configure.sh` output**
~~~
-- The C compiler identification is AppleClang 11.0.0.11000033
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building Production build
-- Performing Test HAVE_FLAG_O3
-- Performing Test HAVE_FLAG_O3 - Success
-- Configuring with C flag '-O3'
-- Configuring with CXX flag '-O3'
-- Performing Test HAVE_FLAG_Wall
-- Performing Test HAVE_FLAG_Wall - Success
-- Configuring with C flag '-Wall'
-- Configuring with CXX flag '-Wall'
-- Performing Test HAVE_FLAG_fexceptions
-- Performing Test HAVE_FLAG_fexceptions - Success
-- Configuring with C flag '-fexceptions'
-- Performing Test HAVE_FLAG_Wno_deprecated
-- Performing Test HAVE_FLAG_Wno_deprecated - Success
-- Configuring with C flag '-Wno-deprecated'
-- Configuring with CXX flag '-Wno-deprecated'
-- Performing Test HAVE_FLAG_Wsuggest_override
-- Performing Test HAVE_FLAG_Wsuggest_override - Failed
-- Performing Test HAVE_FLAG_Wnon_virtual_dtor
-- Performing Test HAVE_FLAG_Wnon_virtual_dtor - Success
-- Configuring with CXX flag '-Wnon-virtual-dtor'
-- Performing Test HAVE_FLAG_Wimplicit_fallthrough
-- Performing Test HAVE_FLAG_Wimplicit_fallthrough - Success
-- Configuring with C flag '-Wimplicit-fallthrough'
-- Configuring with CXX flag '-Wimplicit-fallthrough'
-- Performing Test HAVE_FLAG_Wno_class_memaccess
-- Performing Test HAVE_FLAG_Wno_class_memaccess - Failed
-- Found PythonInterp: /usr/bin/python (found version "2.7.16")
-- Found GMP: /usr/local/include
-- Found GMP libs: /usr/local/lib/libgmp.dylib
-- Performing Test CVC4_NEED_INT64_T_OVERLOADS
-- Performing Test CVC4_NEED_INT64_T_OVERLOADS - Success
-- Performing Test CVC4_NEED_HASH_UINT64_T_OVERLOAD
-- Performing Test CVC4_NEED_HASH_UINT64_T_OVERLOAD - Failed
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for C++ include ext/stdio_filebuf.h
-- Looking for C++ include ext/stdio_filebuf.h - not found
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for ffs
-- Looking for ffs - found
-- Looking for optreset
-- Looking for optreset - found
-- Looking for sigaltstack
-- Looking for sigaltstack - found
-- Looking for strerror_r
-- Looking for strerror_r - found
-- Looking for strtok_r
-- Looking for strtok_r - found
-- Performing Test STRERROR_R_CHAR_P
-- Performing Test STRERROR_R_CHAR_P - Failed
-- Found Git: /usr/bin/git (found version "2.21.0 (Apple Git-122.2)")
-- Looking for antlr3FileStreamNew
-- Looking for antlr3FileStreamNew - found
-- Found ANTLR: /Users/elipol/CVC4/deps/install/bin/antlr3
-- Found ANTLR libs: /Users/elipol/CVC4/deps/install/lib/libantlr3c.a
-- Found Java: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/bin/java (found version "1.8.0.232") found components:  Runtime
CVC4 1.8-prerelease

Build profile        : production

GPL                  : off
Best configuration   : off
Optimized            : on
Optimization level   : off

Assertions           : off
Debug symbols        : off
Debug context mem mgr: off

Dumping              : on
Muzzle               : off
Proofs               : on
Replay               : off
Statistics           : on
Tracing              : off

ASan                 : off
UBSan                : off
TSan                 : off
Coverage (gcov)      : off
Profiling (gprof)    : off
Unit tests           : off
Valgrind             : off

Shared libs          : on
Static binary        : off
Java bindings        : off
Python bindings      : off
Python2              : off
Python3              : off

ABC                  : off
CaDiCaL              : off
CryptoMiniSat        : off
drat2er              : off
GLPK                 : off
LFSC                 : off
MP library           : gmp
Readline             : off
SymFPU               : off


CPPLAGS (-D...)      : NDEBUG CVC4_DUMPING CVC4_PROOF CVC4_STATISTICS_ON
CXXFLAGS             : -O3 -Wall -Wno-deprecated -Wnon-virtual-dtor -Wimplicit-fallthrough
CFLAGS               : -O3 -Wall -fexceptions -Wno-deprecated -Wimplicit-fallthrough

Install prefix       : /usr/local

CVC4 license         : modified BSD

Note that this configuration is NOT built against any GPL'ed libraries, so
it is covered by the (modified) BSD license.  This is, however, not the best
performing configuration of CVC4.  To build against GPL'ed libraries which
improve CVC4's performance, re-configure with '-DENABLE_GPL -DENABLE_BEST'.

Now change to 'build' and type make, followed by make check or make install.

-- Configuring done
-- Generating done
-- Build files have been written to: /Users/elipol/CVC4/build
~~~

**Comment by ajreynol** [comment]
Hi Elizabeth,
We have yet to update to the version 2.0 sygus language by default.  We plan to do this in the next stable release.
For now, you can add `--lang=sygus2` to make CVC4 parse files like the one above.
Do let me know if that works for you.

**Comment by polgreen** [comment]
Ah ok, thanks. That fixes it for the example I gave above, but then I interpret this sentence in SyGuS2 to mean that the example below should also parse.

"The keyword Start, which denoted the starting (non-terminal) symbol of grammars in the previous standard, has been removed. Instead, the first symbol listed in the grammar is assumed to be the starting symbol."

I read this sentence to mean that `I` would be assumed to be the starting symbol.
~~~
(set-logic LIA)
(synth-fun max2 ((x Int) (y Int)) Int
    ((I Int (x y 0 1
             (+ I I) (- I I)
             (ite B I I)))
     (B Bool ((and B B) (or B B) (not B)
              (= I I) (<= I I) (>= I I))))
)

(declare-var x Int)
(declare-var y Int)
(constraint (>= (max2 x y) x))
(constraint (>= (max2 x y) y))
(constraint (or (= x (max2 x y)) (= y (max2 x y))))
(check-synth)
~~~

I think this should be valid sygus 2.0 and not valid sygus 1.0? But
- This parses if I run `cvc4 test.sl` at commit bcaebfa1
- It does not parse if I run `cvc4 --lang=sygus2 test.sl` at commit bcaebfa1. 
- It does not parse running `cvc4 test.sl` at commit a3bcd08da.

**Comment by ajreynol** [comment]
Yes, that is correct above that `I` is the starting symbol. However the above benchmark is neither v1 or v2, so technically it should be rejected.

However, our sygus v1 parser is permissive in that it will assume the first symbol is the start symbol.

I am not sure why this didn't parse as v1 previously, but it does parse in v1 (i.e. no command line options) in current master.

**Comment by polgreen** [comment]
Ah, so the predeclaration is mandatory. ok, thanks. 

I wasn't sure because, although the SyGuS documentation in section 3.4 it says that a grammar G consists of two parts: a predeclaration and a group rule listing, in section 6.1 it just says that a grammar G is specified as a group rule listing. 
3.4:
![image](https://user-images.githubusercontent.com/17899331/76364266-f8d39180-62e1-11ea-9a72-2fb19c728aee.png)
6.1:
![image](https://user-images.githubusercontent.com/17899331/76364205-db062c80-62e1-11ea-800c-34c630786acf.png)

**Comment by ajreynol** [comment]
Ah, good catch, I can update the sygus language standard to fix this (in 6.1, it should be fixed to say "G is specified by a predeclaration and a grouped rule listing").
